### ISSUE TITLE
Small fixes for domain_hashdump

### ISSUE THREAD

**Issue by mubix** [rapid7/metasploit-framework#5645]
cc: @dmaloney-r7 

### Code
- [x] Use a better method of determining if the system is a Domain Controller. NTDS on production systems is 50/50 to be stored on the %SYSTEMDRIVE%. There is a registry key that identifies where the NTDS.DIT is (see next bullet). Finding an NTDS.dit is a good indication that its a Domain Controller, but, so we can remove the 
- [x] Add registry check for where the NTDS.dit is located `HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\services\NTDS\Parameters   Value: "DSA Database file"`
- [ ] Creating a temp directory for the NTDSUTIL method on the system disk will result in disk full errors if the DC is partitioned. It would probably be better to either check this size of the NTDS then make a determination where to put it, or make a directory in the same location of the NTDS to be safe, since it's a good bet the partition it's on has a good amount of space.
- [x] Please put functions in order of use if possible. Reading it is tough scrolling all over the place.
- [ ] Add references, the blog post had a load of them, it'd probably be good for posterity to include them in the module.
- [x] Could you comment on the reasoning behind the preference for the NTDSUTIL method vs just staying with VSS?
- [x] Getting the domain name should be a registry query instead of creating a cmd.exe process or an API call. The registry key is `HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Services\Tcpip\Parameters\Domain` and the API call can be: `GetComputerNameEx` w/ Railgun or Meterpreter itself might even have the function (or should? @OJ ). The other option is if we are already in the NTDS service registry key for finding the NTDS, you can parse the `Machine DN Name` for retrieving the domain name, which is guaranteed to be the one in the NTDS file.
- [ ] The NTDSUTIL method creates a complete copy of not only the NTDS.dit but also the SYSTEM and SAM hives. These should probably be cleaned up the operation is complete. Leaving them in C:\Windows\Temp as it currently stands is not great.
- [x] The UAC and admin check is unnecessary the `is_admin` will result in a `FALSE` if you are behind UAC, so you could say "Current user is not admin or is behind UAC" with just the `is_admin` check.
- [x] If the user is under UAC or not an Admin, the check for the `%SYSTEMDRIVE%\Windows\` directory will fail and say "ACCESS DENIED" so the Admin check should be first and bail if  you aren't an admin.
- [x] Why is there an `RHOST` variable? All of the commands seem to be local so probably de-registering that variable would help with any confusion.
- [x] This also goes for the SMBUser SMBPass and SMBDomain variables, unless they are there for a future addition to allow you to do this remotely some how.
- [x] Why do you copy the domain_name variable to "realm", it doesn't look like you modify it afterwards?
- [x] For the VSS Method the same about the %SYSTEMDRIVE% applies. You won't find the NTDS there always. Plus you can create a VSS copy of a very specific folder (say where the NTDS is) and then destroy the VSS copy afterwards for clean up.
- [x] Why are you moving the VSS copy of the NTDS file? You should be able to parse it directly from the VSS copy as it should act very much like a file on the system. (and again with putting it in WINDIR\Temp you will face space problems.
- [ ] The repair looks awesome, didn't know about that command, however the repair should probably only be performed (as it creates a new process) only if there is trouble parsing the NTDS file.

### Output

- [ ]  For the repair, we probably don't need all of the output of the command except maybe in verbose mode. 
- [ ] I'm getting a odd listing for my password being set to never expire `Password Expires: r` might be concatenating a string? or trying to convert it to a date? (look at the KRBTGT user below)
- [ ] Probably should remove the "Hash History" output if there isn't any for the user
- [ ] Getting some odd tabs, other spacing, and dashes:
```
[+] Administrator (Built-in account for administering the computer/domain)
Administrator:500:aad3b435b51404eeaad3b435b51404ee:88E4D9FABAECF3DEC18DD80905521B29
Password Expires: Monday, January 01, 1601
Last Password Change: 10:38:03 AM Friday, December 05, 2014
Last Logon: 1:57:07 AM Saturday, December 06, 2014
Logon Count: 12
 - Password Never Expires

Hash History:


[+]          Guest (Built-in account for guest access to the computer/domain)
         Guest:501:aad3b435b51404eeaad3b435b51404ee:31d6cfe0d16ae931b73c59d7e0c089c0
         Password Expires: Never
         Last Password Change: 12:00:00 AM Monday, January 01, 1601
         Last Logon: 12:00:00 AM Monday, January 01, 1601
         Logon Count: 0
          - Account Disabled
- Password Never Expires
- No Password Required

         Hash History:


[+] krbtgt (Key Distribution Center Service Account)
krbtgt:502:aad3b435b51404eeaad3b435b51404ee:88E4D9FABAECF3DEC18DD80905521B29
Password Expires: r
Last Password Change: 7:41:38 AM Friday, December 05, 2014
Last Logon: 12:00:00 AM Monday, January 01, 1601
Logon Count: 0
 - Account Disabled

Hash History:
krbtgt:502:aad3b435b51404eeaad3b435b51404ee:88E4D9FABAECF3DEC18DD80905521B29


[+] user ()
user:1103:aad3b435b51404eeaad3b435b51404ee:88E4D9FABAECF3DEC18DD80905521B29
Password Expires: r
Last Password Change: 7:45:16 AM Friday, December 05, 2014
Last Logon: 3:20:13 AM Saturday, December 06, 2014
Logon Count: 34
 - Password Never Expires

Hash History:
user:1103:aad3b435b51404eeaad3b435b51404ee:88E4D9FABAECF3DEC18DD80905521B29
```

If I was to reformat the output it would look like this:
```
[+] Administrator (Built-in account for administering the computer/domain)
         Administrator:500:aad3b435b51404eeaad3b435b51404ee:88E4D9FABAECF3DEC18DD80905521B29
         Password Expires: Monday, January 01, 1601
         Last Password Change: 10:38:03 AM Friday, December 05, 2014
         Last Logon: 1:57:07 AM Saturday, December 06, 2014
         Logon Count: 12
         Account Statuses: Password Never Expires

[+] Guest (Built-in account for guest access to the computer/domain)
         Guest:501:aad3b435b51404eeaad3b435b51404ee:31d6cfe0d16ae931b73c59d7e0c089c0
         Password Expires: Never
         Last Password Change: 12:00:00 AM Monday, January 01, 1601
         Last Logon: 12:00:00 AM Monday, January 01, 1601
         Logon Count: 0
         Account Statuses: Account Disabled, Password Never Expires, No Password Required

[+] krbtgt (Key Distribution Center Service Account)
         krbtgt:502:aad3b435b51404eeaad3b435b51404ee:88E4D9FABAECF3DEC18DD80905521B29
         Password Expires: r
         Last Password Change: 7:41:38 AM Friday, December 05, 2014
         Last Logon: 12:00:00 AM Monday, January 01, 1601
         Logon Count: 0
         Account Statuses: Account Disabled
         Hash History:
           krbtgt:502:aad3b435b51404eeaad3b435b51404ee:88E4D9FABAECF3DEC18DD80905521B29

[+] user ()
         user:1103:aad3b435b51404eeaad3b435b51404ee:88E4D9FABAECF3DEC18DD80905521B29
         Password Expires: r
         Last Password Change: 7:45:16 AM Friday, December 05, 2014
         Last Logon: 3:20:13 AM Saturday, December 06, 2014
         Logon Count: 34
         Account Statuses: Password Never Expires
         Hash History:
           user:1103:aad3b435b51404eeaad3b435b51404ee:88E4D9FABAECF3DEC18DD80905521B29
```

**Comment by OJ** [comment]
I thought we already had the domain via `sysinfo` ?

https://github.com/rapid7/metasploit-framework/blob/e749733eb6118a4b089e288fc81050f76c8db5ed/lib/rex/post/meterpreter/extensions/stdapi/sys/config.rb#L97

**Comment by dmaloney-r7** [comment]
1) Somebody else already pointed out eh Registry key we can use to fin d the NTDS location. We will definitely make that change
2) not so sure about the location to move the copy to, i take your point about partitioning, this bears some more pondering
3) Methods are put in lexical ordering as per both our own and the ruby style guide, this will not change
4) ntdsutil is prefered over straight VSS because it the NTDS database is copied in a proper disconnected state. Using the VSS method will always, in my experience, leave the database in a bad state which then requires the repair operation to be run on it before we can connect to it. ntdsutil leaves the database in a proper state
5) The domain information can now be gotten from sysinfo, so we'll actually be using that going forward. Those PRs were issued independently, and I didn't want the NTDS pull to depend on the other one
6) good call on the is_admin checks, i will make the changes you suggest here
7) stupid mixins, those options got mixed in, I'll add a deregister on them
8) realm is part of the Credential::Core, a Realm represent an authentication realm the cred is valid for, such as a Domain, or in the case of postgres a Datbase, Oracle a SID etc
9) As for moving the VSS copy, it's mainly because i'm not sure the parser will behave correctly with that path, I had issues with UNC paths when doing some early testing
10) the repair is always necessary if the VSS method was used instead of ntdsutil
11) I personally like having all of that output from the repair, but it could be moved to VERBOSE mode if people don'y like it
12) That date is weird, it should probably be the string "Never", but it's getting truncated somehow

@mubix , hope that addresses all of your questions/concerns. i'll go through and make some of the changes as discussed tomorrow and try to get them up in a new PR

**Comment by dmaloney-r7** [comment]
@oj yup, i i put that pr up separately, i didn't want to mix the two at the time as there was already a lot of movement going on with the NTDS stuff. i'll be changing the module to use the sysinfo field

**Comment by mubix** [comment]
1) Cool beans
2) I've run into the C:\ drive being very small a lot and then D:\ being the "data" drive where NTDS.dit would sit. It's been taught to do this for a long time in most trainings on how to set up a domain controller so that backups can just include the "important" non-OS stuff.
3) Never heard of Lexical format, thanks, will try to conform my code to the same going forward
4) Cool, but with the ability to repair it that means there really isn't much difference. I'd be curious to see which method creates more forensicly "loud" logs.
5 - 7) Cool, thanks!
8) I understand that but, technically you don't have to rename the variable for it to be applied to the function as "realm". Is that another ruby style guide preference?
9 & 10) It might make sense to default to VSS, check to see if the repair is needed, if so, fail over to the NTDSUTIL version if it's available from the OS, then doing the copy/repair for VSS if needed. Since one of your main points on the blog post is being quiet and not getting caught.
11) I usually stay with the "if it isn't made by metasploit, it's verbose" style guide ;-)
12) Any idea what would cause the truncation on just the last two users?

Thanks!

**Comment by mubix** [comment]
Another thought, I know this is all going into the database, and everyone expects the database to be running (I rarely have a DB on small hardware like a Pi or a Hak5 Pineapple), but I really like output files, especially when it comes to creds. Any chance of adding an "OUTFILE" or a loot export for all of the creds and history so it's easy to pipe into John.

**Comment by jabra-** [comment]
+1 for JTR output!

**Comment by dmaloney-r7** [comment]
@mubix Lexical ordering == Alphabetical essentially

**Comment by dmaloney-r7** [comment]
@mubix ntdsutil is definitely the superior approach, however it is only available on 2k8 and later boxes. We prefer ntdustil because it is cleaner it's also the way someone is 'supposed' to make copies of NTDS.dit we only fall back to manual VSS when we have to. Also the manual VSS method makes a shadow copy of the entire volume, which is not ideal

**Comment by dmaloney-r7** [comment]
8) domain_name isn't a variable, it's a method that returns a value. iit is then put into a variable called realm which is used in multiple subsequent calls

**Comment by dmaloney-r7** [comment]
@mubix yeah, as i thought, tested parsing directly from the shadowcopy, jet will not attach to it there, it has to be moved onto a real volume

**Comment by dmaloney-r7** [comment]
@mubix also another big difference between ntdsutil and VSS, is that ntdsutil already knows where the file is, whereas with VSS we have to look for it explicitly.

**Comment by dmaloney-r7** [comment]
Most of these issues have been resolved in the pr already, closing this for now. if there are issues that are still big cocnerns please open a new issue =)
